File: android/lib/models/message.dart
// Copyright 2026 Daniel Pelikan
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import 'dart:convert';

/// Protocol version.
const int protocolVersion = 1;

/// Message types supported by the protocol.
enum MessageType {
  text('TEXT'),
  command('COMMAND'),
  heartbeat('HEARTBEAT'),
  ack('ACK'),
  pairReq('PAIR_REQ'),
  pairAck('PAIR_ACK');

  final String value;
  const MessageType(this.value);

  static MessageType fromString(String s) {
    return MessageType.values.firstWhere(
      (e) => e.value == s,
      orElse: () => throw ArgumentError('Unknown message type: $s'),
    );
  }
}

/// Protocol message structure.
class Message {
  final int version;
  final MessageType messageType;
  String payload;
  final int timestamp;
  String checksum;

  Message({
    required this.version,
    required this.messageType,
    required this.payload,
    required this.timestamp,
    this.checksum = '',
  });

  /// Create a new message with automatic timestamp.
  factory Message.create(MessageType type, String payload) {
    return Message(
      version: protocolVersion,
      messageType: type,
      payload: payload,
      timestamp: DateTime.now().millisecondsSinceEpoch,
    );
  }

  /// Create a TEXT message.
  factory Message.text(String content) => Message.create(MessageType.text, content);

  /// Create a COMMAND message.
  factory Message.command(String cmd) => Message.create(MessageType.command, cmd);

  /// Create a HEARTBEAT message.
  factory Message.heartbeat() => Message.create(MessageType.heartbeat, '');

  /// Create an ACK message.
  factory Message.ack(int originalTimestamp) =>
      Message.create(MessageType.ack, originalTimestamp.toString());

  /// Create a PAIR_REQ message.
  factory Message.pairRequest(PairRequestPayload payload) =>
      Message.create(MessageType.pairReq, payload.toJson());

  /// Create a PAIR_ACK message.
  factory Message.pairAck(PairAckPayload payload) =>
      Message.create(MessageType.pairAck, payload.toJson());

  /// Serialize to JSON string with newline delimiter.
  String toJson() {
    final map = {
      'v': version,
      't': messageType.value,
      'p': payload,
      'ts': timestamp,
      'cs': checksum,
    };
    return '${jsonEncode(map)}\n';
  }

  /// Parse from JSON string.
  factory Message.fromJson(String json) {
    final map = jsonDecode(json.trim()) as Map<String, dynamic>;
    return Message(
      version: map['v'] as int,
      messageType: MessageType.fromString(map['t'] as String),
      payload: map['p'] as String,
      timestamp: map['ts'] as int,
      checksum: map['cs'] as String? ?? '',
    );
  }

  /// Check if payload should be encrypted.
  bool get shouldEncrypt {
    return messageType == MessageType.text ||
        messageType == MessageType.command ||
        messageType == MessageType.pairReq ||
        messageType == MessageType.pairAck;
  }
}

/// Pairing request payload.
class PairRequestPayload {
  final String deviceId;
  final String? deviceName;

  PairRequestPayload({
    required this.deviceId,
    this.deviceName,
  });

  String toJson() {
    final map = <String, dynamic>{
      'device_id': deviceId,
    };
    if (deviceName != null) {
      map['device_name'] = deviceName;
    }
    return jsonEncode(map);
  }

  factory PairRequestPayload.fromJson(String json) {
    final map = jsonDecode(json) as Map<String, dynamic>;
    return PairRequestPayload(
      deviceId: map['device_id'] as String,
      deviceName: map['device_name'] as String?,
    );
  }
}

/// Pairing acknowledgment payload.
class PairAckPayload {
  final String deviceId;
  final PairStatus status;
  final String? error;

  PairAckPayload({
    required this.deviceId,
    required this.status,
    this.error,
  });

  factory PairAckPayload.success(String deviceId) => PairAckPayload(
        deviceId: deviceId,
        status: PairStatus.ok,
      );

  factory PairAckPayload.error(String deviceId, String error) => PairAckPayload(
        deviceId: deviceId,
        status: PairStatus.error,
        error: error,
      );

  String toJson() {
    final map = <String, dynamic>{
      'device_id': deviceId,
      'status': status == PairStatus.ok ? 'ok' : 'error',
    };
    if (error != null) {
      map['error'] = error;
    }
    return jsonEncode(map);
  }

  factory PairAckPayload.fromJson(String json) {
    final map = jsonDecode(json) as Map<String, dynamic>;
    return PairAckPayload(
      deviceId: map['device_id'] as String,
      status: map['status'] == 'ok' ? PairStatus.ok : PairStatus.error,
      error: map['error'] as String?,
    );
  }
}

enum PairStatus { ok, error }

File: android/android/app/src/main/java/com/speech2prompt/domain/model/Message.kt
package com.speech2prompt.domain.model

import kotlinx.serialization.KSerializer
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.descriptors.PrimitiveKind
import kotlinx.serialization.descriptors.PrimitiveSerialDescriptor
import kotlinx.serialization.descriptors.SerialDescriptor
import kotlinx.serialization.encoding.Decoder
import kotlinx.serialization.encoding.Encoder
import kotlinx.serialization.json.Json
import java.security.MessageDigest

/**
 * Protocol version for message compatibility
 */
const val PROTOCOL_VERSION = 3

/**
 * JSON configuration for message serialization
 */
val messageJson = Json {
    ignoreUnknownKeys = true
    encodeDefaults = true
    isLenient = true
}

/**
 * Message type enumeration with wire values
 */
@Serializable(with = MessageTypeSerializer::class)
enum class MessageType(val value: String) {
    TEXT("TEXT"),
    WORD("WORD"),  // Single word with sequence number
    COMMAND("COMMAND"),
    HEARTBEAT("HEARTBEAT"),
    ACK("ACK"),
    PAIR_REQ("PAIR_REQ"),
    PAIR_ACK("PAIR_ACK");

    companion object {
        fun fromValue(value: String): MessageType {
            return entries.find { it.value.equals(value, ignoreCase = true) }
                ?: throw IllegalArgumentException("Unknown message type: $value")
        }
    }
}

/**
 * Custom serializer for MessageType to use value field
 */
object MessageTypeSerializer : KSerializer<MessageType> {
    override val descriptor: SerialDescriptor =
        PrimitiveSerialDescriptor("MessageType", PrimitiveKind.STRING)

    override fun serialize(encoder: Encoder, value: MessageType) {
        encoder.encodeString(value.value)
    }

    override fun deserialize(decoder: Decoder): MessageType {
        return MessageType.fromValue(decoder.decodeString())
    }
}

/**
 * Payload for WORD messages - single word with session ID
 */
@Serializable
data class WordPayload(
    @SerialName("word")
    val word: String,
    @SerialName("session")
    val session: String,
    @SerialName("ts")
    val ts: Long = System.currentTimeMillis()
)

/**
 * Protocol message for BLE communication.
 * Matches Flutter's Message class exactly.
 */
@Serializable
data class Message(
    @SerialName("v")
    val version: Int = PROTOCOL_VERSION,

    @SerialName("t")
    val messageType: MessageType,

    @SerialName("p")
    var payload: String,

    @SerialName("ts")
    val timestamp: Long = System.currentTimeMillis(),

    @SerialName("cs")
    var checksum: String = ""
) {
    /**
     * Whether this message type should be encrypted
     */
    val shouldEncrypt: Boolean
        get() = when (messageType) {
            MessageType.PAIR_REQ,
            MessageType.PAIR_ACK,
            MessageType.ACK,
            MessageType.HEARTBEAT -> false
            else -> true
        }

    /**
     * Serialize message to JSON string.
     * 
     * NOTE: For messages requiring checksum verification (TEXT, COMMAND),
     * use signAndToJson(secret) to include cryptographic checksum.
     * Messages without a secret will have an empty checksum.
     */
    fun toJson(): String {
        return messageJson.encodeToString(serializer(), this)
    }
    
    /**
     * Sign the message with the shared secret and serialize to JSON.
     * Uses SHA-256 checksum compatible with desktop implementation.
     * 
     * @param secret The shared secret key (32 bytes)
     * @return JSON string with cryptographic checksum
     */
    fun signAndToJson(secret: ByteArray?): String {
        if (secret != null) {
            checksum = computeChecksum(secret)
        }
        return messageJson.encodeToString(serializer(), this)
    }

    /**
     * Compute SHA-256 checksum of message content.
     * Format: SHA256(version + msgType + payload + timestamp + secret)[0:4] as hex
     * This matches the desktop Rust implementation exactly.
     * 
     * @param secret The shared secret key (32 bytes)
     * @return 8-character hex string (first 4 bytes of SHA-256 hash)
     */
    fun computeChecksum(secret: ByteArray): String {
        val md = MessageDigest.getInstance("SHA-256")
        md.update(version.toString().toByteArray(Charsets.UTF_8))
        md.update(messageType.value.toByteArray(Charsets.UTF_8))
        md.update(payload.toByteArray(Charsets.UTF_8))
        md.update(timestamp.toString().toByteArray(Charsets.UTF_8))
        md.update(secret)
        
        val hash = md.digest()
        // Take first 4 bytes and convert to hex (lowercase)
        return hash.take(4).joinToString("") { byte -> "%02x".format(byte) }
    }

    /**
     * Verify message checksum against a computed checksum.
     * 
     * @param secret The shared secret key (32 bytes)
     * @return true if checksum matches or is empty, false otherwise
     */
    fun verifyChecksum(secret: ByteArray): Boolean {
        if (checksum.isEmpty()) return true
        return checksum.equals(computeChecksum(secret), ignoreCase = true)
    }

    companion object {
        /**
         * Parse message from JSON string
         */
        fun fromJson(json: String): Message {
            return messageJson.decodeFromString(serializer(), json)
        }

        /**
         * Create a text message
         */
        fun text(content: String): Message {
            return Message(
                messageType = MessageType.TEXT,
                payload = content
            )
        }

        /**
         * Create a command message
         */
        fun command(cmd: String): Message {
            return Message(
                messageType = MessageType.COMMAND,
                payload = cmd
            )
        }

        /**
         * Create a heartbeat message
         */
        fun heartbeat(): Message {
            return Message(
                messageType = MessageType.HEARTBEAT,
                payload = ""
            )
        }

        /**
         * Create an acknowledgment message
         */
        fun ack(originalTimestamp: Long): Message {
            return Message(
                messageType = MessageType.ACK,
                payload = originalTimestamp.toString()
            )
        }

        /**
         * Create a pairing request message
         */
        fun pairRequest(payload: PairRequestPayload): Message {
            return Message(
                messageType = MessageType.PAIR_REQ,
                payload = messageJson.encodeToString(PairRequestPayload.serializer(), payload)
            )
        }

        /**
         * Create a pairing acknowledgment message
         */
        fun pairAck(payload: PairAckPayload): Message {
            return Message(
                messageType = MessageType.PAIR_ACK,
                payload = messageJson.encodeToString(PairAckPayload.serializer(), payload)
            )
        }

        /**
         * Create a word message with session ID
         */
        fun word(word: String, session: String): Message {
            val payload = WordPayload(word, session)
            return Message(
                messageType = MessageType.WORD,
                payload = messageJson.encodeToString(WordPayload.serializer(), payload)
            )
        }
    }
}

/**
 * Extension to parse payload as PairRequestPayload
 */
fun Message.parsePairRequestPayload(): PairRequestPayload? {
    if (messageType != MessageType.PAIR_REQ) return null
    return try {
        messageJson.decodeFromString(PairRequestPayload.serializer(), payload)
    } catch (e: Exception) {
        null
    }
}

/**
 * Extension to parse payload as PairAckPayload
 */
fun Message.parsePairAckPayload(): PairAckPayload? {
    if (messageType != MessageType.PAIR_ACK) return null
    return try {
        messageJson.decodeFromString(PairAckPayload.serializer(), payload)
    } catch (e: Exception) {
        null
    }
}
